python/src/opendp/_data.py

```diff
@@ -252,6 +281,86 @@ def object_type(
     return output
 
 
+def onceframe_collect(
+    onceframe
+):
+    r"""Internal function. Collects a DataFrame from a OnceFrame, exhausting the OnceFrame.
+
+    :param onceframe: The queryable holding a LazyFrame.
+    :raises TypeError: if an argument's type differs from the expected type
+    :raises UnknownTypeException: if a type argument fails to parse
+    :raises OpenDPException: packaged error from the core OpenDP library
+    """
+    # No type arguments to standardize.
+    # Convert arguments to c types.
+    c_onceframe = py_to_c(onceframe, c_type=AnyObjectPtr, type_name=AnyQueryable)
+
+    # Call library function.
+    lib_function = lib.opendp_data__onceframe_collect
+    lib_function.argtypes = [AnyObjectPtr]
+    lib_function.restype = FfiResult
+
+    output = c_to_py(unwrap(lib_function(c_onceframe), AnyObjectPtr))
+
+    return output
+
+
+def onceframe_sink_csv(
```

Comment:
If we have a method to return a DataFrame, should that be enough? Is there a concern that going through an intermediate data structure will be too slow or inefficient for large datasets and we need this method and `onceframe_sink_parquet` to make it more direct? If that is the motivation, worry about premature optimization. What determines that these two formats are the ones we support?

Reply:
This is a great comment. That is a motivation, yes, but I also had another hope for OnceFrames that are not yet used by this PR. I think `trait PrivateLogicalPlan` shouldn't emit measurements with type `LogicalPlan`, as this also has the same problem as the `LazyFrame` type. `OnceFrame` better suits this. I was thinking this would be in a future PR.

Arguably we could just make `OnceFrame` internal, and only release data frames from `make_private_lazyframe`. However, grouping key sets can get larger than memory, so permitting these use-cases would be useful. One minor simplification here is that I could remove the auxiliary `sink_*` constructors, and just keep `collect` and the honest-but-curious `lazyframe`.